### PR TITLE
Redefine htlc intercept and processing spec

### PR DESCRIPTION
**Pull request by okjodom**
Improve upon [GatewayLightning](https://github.com/fedimint/fedimint/blob/master/gateway/ln-gateway/proto/gatewaylnrpc.proto#L17:L19) `subscribe_intercept_htlcs` by defining a [bi-directional rpc](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) pattern for subscribing to intercept htlcs and responding after complete htlc processing by the client.

- For every new federation connected, the gateway **sends a request for HTLC interception**:
```
// example
let request = InterceptHtlcsRequest { action: Some(Action::Subscribe(Subscribe { short_channel_id: u64 })) };
let stream = client.intercept_htlcs(request).await?;
```
- For every HTLC intercepted and processed by the gateway:
  - Use this RPC to **send a request to settle the HTLC** if HTLC processing was successful
  ```
  // example
  let request = InterceptHtlcsRequest { action: Some(Action::Settle(Settle { preimage: [] })) };
  let stream = client.intercept_htlcs(request).await?;
  ```
  - Use this RPC to **send a request to cancel the HTLC** if HTLC processing failed, timed-out or was rejected for some other reason
  ```
  // example
  let request = InterceptHtlcsRequest { action: Some(Action::Cancel(Cancel { reason: "" })) };
  let stream = client.intercept_htlcs(request).await?;
  ```

The RPC action types applied here are defined in protobuf message types as `InterceptHtlcsRequest`

### CODE REVIEW

File: gateway/ln-gateway/proto/gatewaylnrpc.proto [hunk 1]
```diff
@@ -38,12 +40,49 @@ message PayInvoiceResponse {
...
+  }
+
+  message Settle {
+    // The preimage for settling an intercepted HTLC
+    bytes preimage = 1;
+  }
+
+  message Cancel {
+    // The reason for the cancellation of an intercepted HTLC
+    string reason = 1;
```
Comment (elsirion) [inline gateway/ln-gateway/proto/gatewaylnrpc.proto:56]: Maybe this should (also) be a LN error code we can send back to the sender?

File: gateway/ln-gateway/proto/gatewaylnrpc.proto [hunk 2]
```diff
@@ -38,12 +40,49 @@ message PayInvoiceResponse {
...
+    ///
+    /// Send this request when the gateway failed or canceled processing of
+    /// intercepted HTLC. GatewayLightning will fail/cancel the intercepted HTLC
+    /// with reason provided.
+    Cancel cancel = 3;
+  }
 }
 
-message SubscribeInterceptHtlcsResponse {
+message InterceptedHtlcResponse {
```
Comment (elsirion) [inline gateway/ln-gateway/proto/gatewaylnrpc.proto:85]: This type's docs seem to talk a lot about invoices, but intercepted HTLCs don't tell you much about the invoice from which it was generated. 

Also the `units` member isn't all that useful imo. Just make it msat? Or is this a LND thing we try to be compatible with?

**Review by elsirion — DISMISSED** [review]

File: gateway/ln-gateway/proto/gatewaylnrpc.proto (hunk 1 above)
Comment (okjodom) [reply gateway/ln-gateway/proto/gatewaylnrpc.proto:56]: 💯 , will formalize into standard LN error codes compatible/translatable across implementations

File: gateway/ln-gateway/proto/gatewaylnrpc.proto (hunk 2 above)
Comment (okjodom) [reply gateway/ln-gateway/proto/gatewaylnrpc.proto:85]: > This type's docs seem to talk a lot about invoices, but intercepted HTLCs don't tell you much about the invoice from which it was generated.
> Also the units member isn't all that useful imo. Just make it msat? Or is this a LND thing we try to be compatible with?

Update to require interceptors send amounts in msat. Both LND and CLN can do this today

**Review by elsirion — APPROVED** [review]

**Comment by okjodom** [comment]
@elsirion, @m1sterc001guy , I'm almost convinced we should have  two RPC functions instead of this proposal for a single bi-directional streaming RPC named `intercept_htlcs`. My new finding when implementing this spec is a client can send multiple messages to the server over a single request, but the RPC api does not clarify the following:
- Can a client initiate multiple htlc subscriptions at the same time?
- Can a client initiate a htlc subscription over the same request which also responds to some earlier intercepted htlcs?

As an alternative, we can have:
- `subscribe_intercept_htlc` for opening new htlc intercept subscriptions
- `complete_processed_htlc` for sending **Settle** or **Cancel** messages to complete intercepted and processed htlcs

**Comment by okjodom** [comment]
Abandoned in favor of #1240

**Pull request closed without merging**